src/box/memtx_tree.cc

```diff
@@ -922,6 +922,8 @@ memtx_tree_index_gc_run(struct memtx_gc_task *task, bool *done)
 			tuple_unref((struct tuple *)res->hint);
 		else
 			tuple_unref(res->tuple);
+		ERROR_INJECT_DOUBLE(ERRINJ_MEMTX_GC_TIMEOUT, inj->dparam > 0,
+				    thread_sleep(inj->dparam));
```

Comment:
Done.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 2a62b412b..f73130c49 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -1450,7 +1450,11 @@ memtx_engine_gc_f(va_list va)
 	while (!fiber_is_cancelled()) {
 		FiberGCChecker gc_check;
 		bool stop;
-		ERROR_INJECT_YIELD(ERRINJ_MEMTX_DELAY_GC);
+		while (errinj(ERRINJ_MEMTX_DELAY_GC, ERRINJ_BOOL)->bparam) {
+			fiber_sleep(0.001);
+			if (fiber_is_cancelled())
+				return 0;
+		}
 		memtx_engine_run_gc(memtx, &stop);
 		if (stop) {
 			fiber_yield_timeout(TIMEOUT_INFINITY);
diff --git a/src/box/memtx_tree.cc b/src/box/memtx_tree.cc
index c408e0185..830beec74 100644
--- a/src/box/memtx_tree.cc
+++ b/src/box/memtx_tree.cc
@@ -922,8 +922,6 @@ memtx_tree_index_gc_run(struct memtx_gc_task *task, bool *done)
 			tuple_unref((struct tuple *)res->hint);
 		else
 			tuple_unref(res->tuple);
-		ERROR_INJECT_DOUBLE(ERRINJ_MEMTX_GC_TIMEOUT, inj->dparam > 0,
-				    thread_sleep(inj->dparam));
 		if (++loops >= YIELD_LOOPS) {
 			*done = false;
 			return;
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index b06a10c44..778310697 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -106,7 +106,6 @@ struct errinj {
 	_(ERRINJ_IPROTO_WRITE_ERROR_DELAY, ERRINJ_BOOL, {.bparam = false})\
 	_(ERRINJ_LOG_ROTATE, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_MEMTX_DELAY_GC, ERRINJ_BOOL, {.bparam = false}) \
-	_(ERRINJ_MEMTX_GC_TIMEOUT, ERRINJ_DOUBLE, {.dparam = 0}) \
 	_(ERRINJ_NETBOX_DISABLE_ID, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_NETBOX_FLIP_FEATURE, ERRINJ_INT, {.iparam = -1}) \
 	_(ERRINJ_NETBOX_IO_DELAY, ERRINJ_BOOL, {.bparam = false}) \
diff --git a/test/box-luatest/shutdown_test.lua b/test/box-luatest/shutdown_test.lua
index fac340ec8..a2e8927b9 100644
--- a/test/box-luatest/shutdown_test.lua
+++ b/test/box-luatest/shutdown_test.lua
@@ -173,60 +173,20 @@ g.test_shutdown_of_hanging_client_fiber = function(cg)
              {filename = log}))
 end
 
--- Test we don't leak tuples if shutdown during active memtx gc.
--- We do memory leak check only for ASAN builds. ERRINJ_MEMTX_GC_TIMEOUT
--- and number of tuples inserted are tuned so that both number of yields
--- of memtx gc fiber and time required to free tuples is large enough
--- so that we cover case when not all tuples are freed at the time
--- memtx shutdown started.
+-- Let's test we free tuples on index drop and immediate shutdown, before
+-- memtx gc fiber have a chance to free all tuples. That is we take care
+-- too free tuples during shutdown. The actual test is done by LSAN.
+-- This way we make sure other tests in similar situation are not flaky
+-- under ASAN.
 g.test_shutdown_memtx_gc_free_tuples = function(cg)
     t.tarantool.skip_if_not_debug()
     cg.server:exec(function()
         local fiber = require('fiber')
         box.schema.create_space('test')
         box.space.test:create_index('pk')
-        box.error.injection.set('ERRINJ_MEMTX_GC_TIMEOUT', 0.001)
-        fiber.set_slice(100)
-        box.begin()
-        for i=1,10000 do
-            box.space.test:insert{i}
-        end
-        box.commit()
+        box.space.test:insert{1}
+        box.error.injection.set('ERRINJ_MEMTX_DELAY_GC', true)
         box.space.test.index.pk:drop()
     end)
     server:stop()
 end
-
-local g_index = t.group('index', {{index_type = 'TREE'}, {index_type = 'HASH'}})
-
-g_index.before_each(function(cg)
-    cg.server = server:new()
-    cg.server:start()
-end)
-
-g_index.after_each(function(cg)
-    if cg.server ~= nil then
-        cg.server:drop()
-    end
-end)
-
--- Test for non ASAN builds we shutdown fast ie we don't wait for all tuples to
--- be freed.
-g_index.test_shutdown_memtx_gc_cancellable = function(cg)
-    t.tarantool.skip_if_not_debug()
-    t.skip_if(tarantool.build.asan, 'non ASAN build required')
-    cg.server:exec(function(index_type)
-        local fiber = require('fiber')
-        box.schema.create_space('test')
-        box.space.test:create_index('pk', {type = index_type})
-        box.error.injection.set('ERRINJ_MEMTX_GC_TIMEOUT', 0.01)
-        fiber.set_slice(100)
-        box.begin()
-        for i=1,10000 do
-            box.space.test:insert{i}
-        end
-        box.commit()
-        box.space.test.index.pk:drop()
-    end, {cg.params.index_type})
-    test_no_hang_on_shutdown(cg.server)
-end
diff --git a/test/box/errinj.result b/test/box/errinj.result
index e89fd0354..df5a3f06f 100644
--- a/test/box/errinj.result
+++ b/test/box/errinj.result
@@ -79,7 +79,6 @@ evals
   - ERRINJ_IPROTO_WRITE_ERROR_DELAY: false
   - ERRINJ_LOG_ROTATE: false
   - ERRINJ_MEMTX_DELAY_GC: false
-  - ERRINJ_MEMTX_GC_TIMEOUT: 0
   - ERRINJ_NETBOX_DISABLE_ID: false
   - ERRINJ_NETBOX_FLIP_FEATURE: -1
   - ERRINJ_NETBOX_IO_DELAY: false

```

</details>